air/src/execution_step/instructions/canon_stream_map_scalar.rs

```diff
@@ -0,0 +1,101 @@
+use super::TraceHandler;
+use crate::execution_step::boxed_value::JValuable;
+use crate::execution_step::instructions::canon::CanonEpilogClosure;
+use crate::execution_step::instructions::canon::StreamWithSerializedView;
+use crate::execution_step::CanonResultAggregate;
+use crate::execution_step::Stream;
+use crate::execution_step::ValueAggregate;
+use crate::log_instruction;
+use crate::trace_to_exec_err;
+use crate::UncatchableError;
+
+use air_interpreter_data::CanonResult;
+use air_parser::ast;
+use air_parser::AirPos;
+use air_trace_handler::merger::MergerCanonResult;
+
+use std::borrow::Cow;
+use std::rc::Rc;
+
+impl<'i> super::ExecutableInstruction<'i> for ast::CanonStreamMapScalar<'i> {
+    #[tracing::instrument(level = "debug", skip(exec_ctx, trace_ctx))]
+    fn execute(&self, exec_ctx: &mut ExecutionCtx<'i>, trace_ctx: &mut TraceHandler) -> ExecutionResult<()> {
+        log_instruction!(call, exec_ctx, trace_ctx);
+        let canon_result = trace_to_exec_err!(trace_ctx.meet_canon_start(), self)?;
+
+        let epilog: &CanonEpilogClosure<'_> = &|stream_with_positions: StreamWithSerializedView,
+                                                exec_ctx: &mut ExecutionCtx<'_>,
+                                                trace_ctx: &mut TraceHandler|
+         -> ExecutionResult<()> {
+            let StreamWithSerializedView {
+                canon_stream,
+                canon_result_cid,
+            } = stream_with_positions;
+
+            let value = JValuable::as_jvalue(&&canon_stream).into_owned();
+            let tetraplet = canon_stream.tetraplet().clone();
+            let position = trace_ctx.trace_pos().map_err(UncatchableError::from)?;
+            let value = CanonResultAggregate::new(
+                Rc::new(value),
+                tetraplet.peer_pk.as_str().into(),
+                &tetraplet.json_path,
+                position,
+            );
+            let result = ValueAggregate::from_canon_result(value, canon_result_cid.clone());
+
+            exec_ctx.scalars.set_scalar_value(self.scalar.name, result)?;
+            trace_ctx.meet_canon_end(CanonResult::new(canon_result_cid));
+            Ok(())
+        };
+
+        match canon_result {
+            MergerCanonResult::CanonResult(canon_result_cid) => {
+                handle_seen_canon(epilog, canon_result_cid, exec_ctx, trace_ctx)
+            }
+            MergerCanonResult::Empty => {
+                let get_stream_or_default: Box<GetStreamClosure<'_>> =
+                    get_stream_or_default_function(self.stream_map.name, self.stream_map.position);
+                handle_unseen_canon(epilog, &get_stream_or_default, &self.peer_id, exec_ctx, trace_ctx)
+            }
+        }
+    }
+}
+
+/// The resulting closure gets underlying stream from a StreamMap in a context
+/// or returns a default empty stream,
+/// it is crucial for deterministic behaviour, for more info see
+/// github.com/fluencelabs/aquavm/issues/346
```

Comment:
Let's add a dot to the end of this comment according to our [standard](https://github.com/fluencelabs/aquavm/blob/master/docs/developer-notes.md#coding-style).